packages/angular_devkit/core/src/virtual-fs/host/memory.ts

```diff
@@ -0,0 +1,200 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { Observable } from 'rxjs/Observable';
+import { Subject } from 'rxjs/Subject';
+import { ArrayObservable } from 'rxjs/observable/ArrayObservable';
+import { _throw } from 'rxjs/observable/throw';
+import {
+  FileAlreadyExistException,
+  FileDoesNotExistException,
+  PathIsDirectoryException,
+  PathIsFileException,
+} from '../../exception/exception';
+import {
+  NormalizedSep,
+  Path,
+  PathFragment,
+  dirname,
+  join,
+  split,
+} from '../path';
+import {
+  FileBuffer,
+  Host,
+  HostCapabilities,
+  HostWatchEvent,
+  HostWatchEventType,
+  HostWatchOptions,
+  Stats,
+} from './interface';
+
+
+export class SimpleMemoryHost implements Host<{}> {
+  private _cache = new Map<Path, FileBuffer>();
+  private _watchers = new Map<Path, [HostWatchOptions, Subject<HostWatchEvent>][]>();
+
+  protected _isDir(path: Path) {
+    for (const p of this._cache.keys()) {
+      if (p.startsWith(path + NormalizedSep)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  protected _updateWatchers(path: Path, type: HostWatchEventType) {
+    const time = new Date();
+    let currentPath = path;
+    let parent: Path | null = null;
+
+    const maybeWatcher = this._watchers.get(currentPath);
+    if (maybeWatcher) {
+      maybeWatcher.forEach(watcher => {
+        const [options, subject] = watcher;
+        subject.next({ path, time, type });
+
+        if (!options.persistent && type == HostWatchEventType.Deleted) {
+          subject.complete();
+          this._watchers.delete(currentPath);
+        }
+      });
+    }
+
+    do {
+      currentPath = parent !== null ? parent : currentPath;
+      parent = dirname(currentPath);
+
+      const maybeWatcher = this._watchers.get(currentPath);
+      if (maybeWatcher) {
+        maybeWatcher.forEach(watcher => {
+          const [options, subject] = watcher;
+          if (!options.recursive) {
+            return;
+          }
+          subject.next({ path, time, type });
+
+          if (!options.persistent && type == HostWatchEventType.Deleted) {
+            subject.complete();
+            this._watchers.delete(currentPath);
+          }
+        });
+      }
+    } while (parent != currentPath);
```

Comment:
Is notifying the parents always necessary? I can imagine foregoing this functionality for performance reasons. 

Maybe it could be an option in `HostWatchOptions`? Not sure all hosts would support it.

Reply:
Yes, because some watchers watch all subdirectories. I've added a check if there are no watchers to skip this.